buildMozillaMach: add options to disable signing requirement and to enable addon sideloading

Infinidoge opened the pull request:
###### Description of changes

When I swapped over to `firefox-devedition` to try and install an unsigned extension from a file, I discovered that the signing requirement wasn't actually properly disabled in the devedition build.

These changes fix that by setting and exporting the environment variable in the `preConfigure` phase (as opposed to setting the env variable for the derivation), toggled using a new argument to the common Firefox builder.

Accordingly, I adjusted `firefox-devedition` and `librewolf` to make use of the new argument, and made the wrapper check for the argument instead of the environment variable. 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Infinidoge committed:
buildMozillaMach: support disabling extension signing enforcement

Previously, derivations specified the environment variable themselves,
which did not actually disable signing enforcement.

mweinelt commented on pkgs/applications/networking/browsers/firefox/wrapper.nix:
```diff
@@ -102,14 +102,11 @@ let
         throw "Firefox addon name needs to be unique"
-      else if requiresSigning && !lib.hasSuffix "esr" browser.name then
+      else if browser.requireSigning && !lib.hasSuffix "esr" browser.name then
         throw "Nix addons are only supported without signature enforcement (eg. Firefox ESR)"
```
This won't hold true anymore with the end of life of the 102 ESR series.

```suggestion
        throw "Nix addons are only supported without signature enforcement (Librewolf or Firefox Developer Edition)"
```

Infinidoge replied:
Oh they're finally changing that? Huh. We should probably add some additional checks for that then.

Infinidoge replied:
Updated the text, added `eg.` in front of your wording for consistency. For my own reference, was this change announced anywhere? I checked a few changelogs for ESR and didn't see any mention of addon signing changes.

mweinelt commented on pkgs/applications/networking/browsers/firefox/wrapper.nix:
```diff
@@ -102,15 +102,12 @@ let
         throw "Firefox addon name needs to be unique"
-      else if requiresSigning && !lib.hasSuffix "esr" browser.name then
-        throw "Nix addons are only supported without signature enforcement (eg. Firefox ESR)"
+      else if browser.requireSigning && !lib.hasSuffix "esr" browser.name then
+        throw "Nix addons are only supported without signature enforcement (eg. LibreWolf or Firefox Developer Edition)"
```
```suggestion
      else if browser.requireSigning then
        throw "Nix addons are only supported without signature enforcement (eg. LibreWolf or Firefox Developer Edition)"
```

There is no ESR release which allows this anymore.

Infinidoge replied:
Hmm. The unfortunate consequence of another option is that (to my knowledge) nothing in Nixpkgs would work with that part of the wrapper by default, thus requiring compiling Firefox yourself.

Would it be reasonable/possible to add those variants to be built by hydra and cached?

Infinidoge replied:
Double checked LibreWolf and they allow addon sideloading by default, so that will at least be built by default. A version of Firefox Developer Edition compiled that allows addon sideloading would be nice, though.

Infinidoge committed:
firefox: correct capitalization in extension check error

Infinidoge commented:
Separated addon sideloading into its own option, which is enabled for only LibreWolf by default, and is now what is checked by the wrapper builder. I also made a small fix to the extension ID check error message.

Would it be possible to have versions of Firefox compiled to allow addon sideloading that are cached, so people don't need to compile Firefox themself?

jopejoe1 commented on pkgs/applications/networking/browsers/firefox/wrapper.nix:
```diff
@@ -102,18 +102,15 @@ let
         throw "Firefox addon name needs to be unique"
-      else if requiresSigning && !lib.hasSuffix "esr" browser.name then
-        throw "Nix addons are only supported without signature enforcement (eg. Firefox ESR)"
+      else if browser.requireSigning && browser.allowAddonSideload then
```
This should throw when either requireSigning is true or allowAddonSideload is false currently is only throws when both are true

```suggestion
      else if browser.requireSigning || !browser.allowAddonSideload then
```

Infinidoge replied:
Good catch, I don't think I would have checked and realized that. Fixed.

jopejoe1 reviewed (approved):
Result of `nixpkgs-review pr 244591` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>firefox-devedition-unwrapped (firefoxPackages.firefox-devedition)</li>
    <li>firefox-devedition-unwrapped.debug (firefoxPackages.firefox-devedition.debug)</li>
    <li>firefox-devedition-unwrapped.symbols (firefoxPackages.firefox-devedition.symbols)</li>
    <li>librewolf-unwrapped</li>
    <li>librewolf-unwrapped.debug</li>
  </ul>
</details>

mweinelt commented on pkgs/applications/networking/browsers/firefox/wrapper.nix:
```diff
@@ -102,18 +102,15 @@ let
-      else if requiresSigning && !lib.hasSuffix "esr" browser.name then
-        throw "Nix addons are only supported without signature enforcement (eg. Firefox ESR)"
+      else if browser.requireSigning || !browser.allowAddonSideload then
+        throw "Nix addons are only supported with signature enforcement disabled and addon sideloading enabled (eg. LibreWolf)"
```
Is `Nix addons` a quirky term for `Nixpkgs packaged Firefox extensions`?

Infinidoge replied:
Probably more accurately a quirky term for "Addons given to Firefox through the wrapper", which by extension is "Firefox extensions packaged with Nix"

mweinelt replied:
I think that should be clarified at some point, but not necessarily in this changeset.

mweinelt commented:
Thanks for working on this. Before merging, I'd like to see the commits split up into roughly the following:

- buildMozillaMach: add support for addon sideloading
  - sideloading is the term for silently installing extensions w/o user confirmation, as required for when users want to manage extensions declaratively.
- buildMozillaMach: add support for disabling extension signing requirements
  - this should include the refactor on librewolf and firefox-devedition
- librewolf: allow addon sideloading
  - a separate commit, since this is not a refactor but a change in behavior
- firefox: consider sideloading capability in extension handling

Feel free to explain the reasons for the relevant changes in each commit message, if you think they would be helpful.

Infinidoge commented:
Alright, will do when I get the chance. 

Technically LibreWolf's change isn't a change in behavior since LibreWolf's build scripts include that build flag, but it makes sense to be a separate commit anyways so I'll note that there.

Infinidoge committed:
librewolf: allow addon sideloading

LibreWolf's build configuration[1] already disables addon sideloading,
so this isn't a change in behavior.

[1] https://codeberg.org/librewolf/source/src/commit/e5762e9d5badf50bbc72f578d01111ed6e92dd74/assets/mozconfig.new#L3

Infinidoge committed:
buildMozillaMach: support addon sideloading

Infinidoge committed:
firefox: consider sideloading capability in extension handling

tnichols217 commented:
> Separated addon sideloading into its own option, which is enabled for only LibreWolf by default, and is now what is checked by the wrapper builder. I also made a small fix to the extension ID check error message.
> 
> Would it be possible to have versions of Firefox compiled to allow addon sideloading that are cached, so people don't need to compile Firefox themself?

Yes please, since this merge I haven't been able to correctly use wrapFirefox even with librewolfUnwrapped as it will keep asking for a browser with signature enforcement disabled and sideloading enabled. Alternatively, would be nice to have precompiled versions of firefox with sideloading enabled that would work with wrapFirefox nicely.

Infinidoge commented:
Hmm, LibreWolf *should* work with `wrapFirefox`, as it's the only one with both disabled.

tnichols217 commented:
Ahh, just found out that the new package updates havent been merged into nixos-unstable yet, but the new package checking in wrapfirefox has, so the default builds of librewolf and firefox-devedition-unwrapped both dont actually build with the new flags yet.

Infinidoge commented:
How... How did that manage to happen. Wild. 

Dev edition doesn't allow sideloading by default, so it won't work with `wrapFirefox` for extensions as is. A PR to add a sideloading-supported version of dev edition would be a good idea.

tnichols217 commented:
Nevermind, I have checked again and the new packages _are_ in nixos-unstable (my bad) but I believe that your "browser.requireSigning" actually checks the variable set in the package build
https://github.com/NixOS/nixpkgs/blob/19cbff58383a4ae384dea4d1d0c823d72b49d614/pkgs/applications/networking/browsers/firefox/wrapper.nix#L122

I noticed all the other mentions to "browser." in the file do not reference any of the inputs to the build. Additionally, when trying to build my own edition with extensions it still errors out on particularly the "browser.requireSigning" even though it should be true for the devedition. When i merge the pkgs.firefox-devedition-unwrapped with { requireSigning = true; } it will only go on to error out on the sideloading condition, leading me to believe that the requireSigning condition does not actually check the config specified in the build

Infinidoge commented:
You want `requireSigning = false;`, because it being true is supposed to cause it to error.

tnichols217 commented:
i meant false, typo in the comment

Infinidoge commented:
The issue seems to be that I forgot to passthru `requireSigning` and `allowAddonSideload`, so when checked they are both null/falsey. https://github.com/Infinidoge/nixpkgs/commit/759a7cc50e170a1137bb172559b961fa657ecb8a should fix it, testing locally now.

tnichols217 commented:
Ah, yeah that would make sense. I also completely forgot to even check that.

Infinidoge commented:
Today I Learned `buildMozillaMach` isn't called with `callPackage`, and instead imported and then `callPackage`'d. How I set up the signing/addon sideloading setup prevented it from being overrided on the unwrapped package. Going to fix that too.

Infinidoge commented:
Made https://github.com/Infinidoge/nixpkgs/commit/491072e797f6d28960f182d45125f1f93b9522ce to fix that. The two combined should put everything in order. Currently waiting for Firefox to finish compiling to test it.

Infinidoge commented:
Firefox decided it wasn't going to compile properly, so I'm just going to make the PR and let OfBorg do it and hope for the best.

tnichols217 commented:
Weird, but your commits look like they should fix the issue, I'll try building your patch later.
